File: dotnet/dpworkshop/workshop1/builders/DatasetBuilder.cs
using System;
using workshop1.models;
using workshop1.validators;

namespace workshop1.builders
{
    public class DatasetBuilder
    {
        private string _content = "";
        private string _description;
        private string _author = "Unknown Author";
        private Func<string, bool> _validator = DataValidators.CsvValidator;

        public static DatasetBuilder Builder()
        {
            return new DatasetBuilder();
        }

        public DatasetBuilder Description(string description)
        {
            _description = description;
            return this;
        }

        public DatasetBuilder Author(string author)
        {
            _author = author;
            return this;
        }

        public DatasetBuilder Validator(Func<string, bool> validator)
        {
            _validator = validator;
            return this;
        }

        public DatasetBuilder AddRow(string row)
        {
            _content += row + "\n";
            return this;
        }

        public Dataset Build()
        {
            CheckReadyToBuild();
            return new Dataset(_content.TrimEnd(), _description, _author, _validator);
        }

        private void CheckReadyToBuild()
        {
            if (_content == "")
            {
                throw new ArgumentException("Content must be set");
            }

            if (_description == null)
            {
                throw new ArgumentException("Description must be set");
            }
        }
    }
}

File: java/dpworkshop/workshop2/src/uk/gov/ukho/dpworkshops/two/builders/DatasetBuilder.java
package uk.gov.ukho.dpworkshops.two.builders;

import uk.gov.ukho.dpworkshops.two.models.Dataset;
import uk.gov.ukho.dpworkshops.two.validators.DataValidators;

import java.util.function.Function;

public class DatasetBuilder {
    private String content = "";
    private String description = null;
    private String author = "Unknown Author";
    private Function<String, Boolean> validator = DataValidators::csvValidator;

    public static DatasetBuilder builder() {
        return new DatasetBuilder();
    }

    public DatasetBuilder description(final String description) {
        this.description = description;
        return this;
    }

    public DatasetBuilder author(final String author) {
        this.author = author;
        return this;
    }

    public DatasetBuilder validator(final Function<String, Boolean> validator) {
        this.validator = validator;
        return this;
    }

    public DatasetBuilder addRow(final String row) {
        this.content += row + "\n";
        return this;
    }

    public Dataset build() {
        checkReadyToBuild();
        return new Dataset(content.stripTrailing(), description, author, validator);
    }

    private void checkReadyToBuild() {
        if (content.equals("")) {
            throw new IllegalArgumentException("content must be set");
        }

        if (description == null) {
            throw new IllegalArgumentException("description must be set");
        }
    }
}
